### PR TITLE
Remove nested case statements

### PR DESCRIPTION
**Pull request by vihu**
The extreme deeply nested cases haunt me at night, this attempts to fix those via "railway" mode

### CODE REVIEW

File: src/blockchain_utils.erl [hunk 1]
```diff
@@ -213,6 +214,19 @@ icdf_select(PopulationList, Rnd) ->
     Sum = lists:sum([Weight || {_Node, Weight} <- PopulationList]),
     icdf_select(PopulationList, normalize_float(Rnd * Sum), normalize_float(Rnd * Sum)).
 
+-spec railway([{term(), fun()}], term()) -> {ok, term()} | {error, term(), term(), term()}.
+railway([], Data) ->
```
Comment (madninja) [inline src/blockchain_utils.erl]: Add some documentation for this function

**Comment by evanmcc** [comment]
I don't like this stuff either, but this function seems hard to use and a lot of ceremony.  maybe just move it to something like the consensus group transaction?  see https://github.com/helium/blockchain-core/blob/master/src/transactions/v1/blockchain_txn_consensus_group_v1.erl#L116

File: src/blockchain_utils.erl (hunk 1 above)
Comment (vihu) [reply src/blockchain_utils.erl]: Added here: 7f51b08

**Comment by vihu** [comment]
> I don't like this stuff either, but this function seems hard to use and a lot of ceremony. maybe just move it to something like the consensus group transaction? see https://github.com/helium/blockchain-core/blob/master/src/transactions/v1/blockchain_txn_consensus_group_v1.erl#L116

lol, it's not _that_ hard https://github.com/helium/blockchain-core/pull/387/files#diff-83068e8683a5f6cb07beba4d7e433628R278. I borrowed heavily from the "railway" oriented approach, where it would short circuit if any subsequent step blows up. But yeah, if you hate this, we don't need to get it in here :)

**Comment by Vagabond** [comment]
This also seems related to https://github.com/erlang/eep/blob/master/eeps/eep-0049.md

File: src/blockchain_utils.erl [hunk 2]
```diff
@@ -213,6 +214,36 @@ icdf_select(PopulationList, Rnd) ->
...
+%%
+%% Refer to the eunit below to get a better understanding.
+%% @end
+%%-------------------------------------------------------------------
+-spec railway([{term(), fun()}], term()) -> {ok, term()} | {error, term(), term(), term()}.
+railway([], Data) ->
+    {ok, Data};
+railway([{Tag, Fun} | Rest], Data) ->
+    case Fun(Data) of
+        {ok, NewData} ->
```
Comment (Vagabond) [inline src/blockchain_utils.erl]: Would suggest an 'ok' return that passes `Data` on to the next function without changes (but indicates the validation succeeded).

File: src/blockchain_utils.erl [hunk 3]
```diff
@@ -213,6 +214,36 @@ icdf_select(PopulationList, Rnd) ->
...
+            railway(Rest, NewData);
+        {error, Reason} ->
+            {error, Tag, Reason, Data};
```
Comment (Vagabond) [inline src/blockchain_utils.erl]: Why do we return the Data in the error cases?

File: src/blockchain_utils.erl (hunk 2 above)
Comment (vihu) [reply src/blockchain_utils.erl]: Added here: cddd354

File: src/blockchain_utils.erl (hunk 3 above)
Comment (vihu) [reply src/blockchain_utils.erl]: Mostly just for inspection

**Comment by vihu** [comment]
This was just an experiment, closing.

**Pull request closed without merging**